opal/mca/accelerator/rocm/accelerator_rocm_module.c

```diff
@@ -0,0 +1,623 @@
+                                                  size_t *size);
+
+static size_t mca_accelerator_rocm_get_ipc_handle_size(void);
+static int mca_accelerator_rocm_get_ipc_handle(void *dev_ptr, void **handle);
+static int mca_accelerator_rocm_open_ipc_handle(void *handle,  void **dev_ptr);
+static int mca_accelerator_rocm_close_ipc_handle(void *dev_ptr);
+
+static int mca_accelerator_rocm_host_register(void *ptr, size_t size);
+static int mca_accelerator_rocm_host_unregister(void *ptr);
+
+static int mca_accelerator_rocm_get_device(int *dev_id);
+static int mca_accelerator_rocm_device_can_access_peer( int *access, int dev1, int dev2);
+
+
+opal_accelerator_base_module_t opal_accelerator_rocm_module =
+{
+    mca_accelerator_rocm_check_addr,
+
+    mca_accelerator_rocm_create_stream,
+    mca_accelerator_rocm_destroy_stream,
+    mca_accelerator_rocm_synchronize_stream,
+
+    mca_accelerator_rocm_create_event,
+    mca_accelerator_rocm_destroy_event,
+    mca_accelerator_rocm_record_event,
+    mca_accelerator_rocm_query_event,
+    mca_accelerator_rocm_synchronize_event,
+
+    mca_accelerator_rocm_memcpy_async,
+    mca_accelerator_rocm_memcpy,
+    mca_accelerator_rocm_matrix_memcpy,
+    mca_accelerator_rocm_memmove,
+    mca_accelerator_rocm_malloc,
+    mca_accelerator_rocm_free,
+    mca_accelerator_rocm_get_address_range,
+
+    mca_accelerator_rocm_get_ipc_handle_size,
+    mca_accelerator_rocm_get_ipc_handle,
+    mca_accelerator_rocm_open_ipc_handle,
+    mca_accelerator_rocm_close_ipc_handle,
+
+    mca_accelerator_rocm_host_register,
+    mca_accelerator_rocm_host_unregister,
+
+    mca_accelerator_rocm_get_device,
+    mca_accelerator_rocm_device_can_access_peer
+};
+
+
+static int mca_accelerator_rocm_check_addr (const void *addr, uint64_t *flags)
+{
+    int ret = -1;
+    hipPointerAttribute_t srcAttr;
+    hipError_t err;
+
+    if (NULL == addr || NULL == flags) {
+        return OPAL_ERR_BAD_PARAM;
+    }
+
+    *flags = 0;
+    err = HIP_FUNCS.hipPointerGetAttributes(&srcAttr, addr);
+    if (hipSuccess == err) {
+        ret = 0;
+        if (hipMemoryTypeDevice == srcAttr.memoryType) {
+            //We might want to set additional flags in a later iteration.
+            //*flags |= MCA_ACCELERATOR_FLAGS_HOST_LDSTR;
+            //*flags |= MCA_ACCELERATOR_FLAGS_HOST_ATOMICS;
+            ret = 1;
+        } else if (hipMemoryTypeUnified == srcAttr.memoryType) {
+            *flags |= MCA_ACCELERATOR_FLAGS_UNIFIED_MEMORY;
+            //*flags |= MCA_ACCELERATOR_FLAGS_HOST_LDSTR;
+            //*flags |= MCA_ACCELERATOR_FLAGS_HOST_ATOMICS;
+            ret = 1;
+        }
+    }
+
+    return ret;
+}
+
+static int mca_accelerator_rocm_create_stream(opal_accelerator_stream_t *stream)
+{
+    if (NULL == stream) {
+        return OPAL_ERR_BAD_PARAM;
+    }
+
+    stream->stream = (hipStream_t *)malloc(sizeof(hipStream_t));
+    if (NULL == stream->stream) {
+        free(stream);
```

Comment:
I don't think that the stream should be freed here (anymore?)

Reply:
And below

Reply:
This was a bug, but I changed the API as well so fixed these alongside.